src/algorithm/minkowski/epa/epa3d.rs

```diff
@@ -0,0 +1,357 @@
+        &self,
+        mut simplex: &mut Vec<SupportPoint<Point3<S>>>,
+        left: &SL,
+        left_transform: &TL,
+        right: &SR,
+        right_transform: &TR,
+    ) -> Option<Contact<Point3<S>>>
+    where
+        SL: SupportFunction<Point = Self::Point>,
+        SR: SupportFunction<Point = Self::Point>,
+        TL: Transform<Self::Point>,
+        TR: Transform<Self::Point>,
+    {
+        if simplex.len() < 4 {
+            return None;
+        }
+        let tolerance = NumCast::from(EPA_TOLERANCE).unwrap();
+        let mut polytope = Polytope::new(&mut simplex);
+        let mut i = 1;
+        loop {
+            let p = {
+                let face = polytope.closest_face_to_origin();
+                let p = SupportPoint::from_minkowski(
+                    left,
+                    left_transform,
+                    right,
+                    right_transform,
+                    &face.normal,
+                );
+                let d = p.v.dot(face.normal);
+                if d - face.distance < tolerance || i >= MAX_ITERATIONS {
+                    return contact(&polytope, face);
+                }
+                p
+            };
+            polytope.add(p);
+            i += 1;
+        }
+    }
+
+    fn new() -> Self {
+        Self {
+            m: marker::PhantomData,
+        }
+    }
+}
+
+#[inline]
+fn contact<S>(polytope: &Polytope<S>, face: &Face<S>) -> Option<Contact<Point3<S>>>
+where
+    S: BaseFloat,
+{
+    Some(Contact::new_with_point(
+        CollisionStrategy::FullResolution,
+        face.normal.clone(), // negate ?
+        face.distance,
+        point(polytope, face),
+    ))
+}
+
+/// This function returns the contact point in world space coordinates on shape A.
+///
+/// Compute the closest point to the origin on the given simplex face, then use that to interpolate
+/// the support points coming from the A shape.
+fn point<S>(polytope: &Polytope<S>, face: &Face<S>) -> Point3<S>
+where
+    S: BaseFloat,
+{
+    let (u, v, w) = barycentric_vector(
+        face.normal * face.distance,
+        polytope.vertices[face.vertices[0]].v,
+        polytope.vertices[face.vertices[1]].v,
+        polytope.vertices[face.vertices[2]].v,
+    );
+
+    polytope.vertices[face.vertices[0]].sup_a * u
+        + polytope.vertices[face.vertices[1]].sup_a.to_vec() * v
+        + polytope.vertices[face.vertices[2]].sup_a.to_vec() * w
+}
+
+#[derive(Debug)]
+struct Polytope<'a, S: 'a>
+where
+    S: BaseFloat,
+{
+    vertices: &'a mut Vec<SupportPoint<Point3<S>>>,
+    faces: Vec<Face<S>>,
+}
+
+impl<'a, S: 'a> Polytope<'a, S>
+where
+    S: BaseFloat,
+{
+    pub fn new(simplex: &'a mut Vec<SupportPoint<Point3<S>>>) -> Self {
+        let faces = Face::new(simplex);
+        Self {
+            vertices: simplex,
+            faces,
+        }
+    }
+
+    pub fn closest_face_to_origin(&'a self) -> &'a Face<S> {
+        let mut face = &self.faces[0];
+        for f in self.faces[1..].iter() {
```

Comment:
could use [min_by_key](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.min_by_key)

Reply:
Doesn't work because float do not implement Ord.

Reply:
And I felt that `self.faces.iter().min_by(|a, b| a.distance.partial_cmp(&b.distance).unwrap()).unwrap()
` wasn't much better.